README.md
![logo-nodezoo][Logo]

# nodezoo-coveralls
[![Npm][BadgeNpm]][Npm]
[![Travis][BadgeTravis]][Travis]
[![Coveralls][BadgeCoveralls]][Coveralls]
[![Gitter][BadgeGitter]][Gitter]

- __Lead:__ [Mihai Dima][Lead]
- __Sponsor:__ [nearForm][Sponsor]

A microservice that provides Coveralls data for [NodeZoo org][]. This microservice depends
on [Coveralls][] and the NPM registry but also caches retrieved data to reduce load on both
public registries.

Please see the [Nodezoo System][System] for more details.

If you're using this microservice, and need help, you can:

- Post a [github issue][Issue],
- Tweet to [@nodezoo][Twitter],
- Ask on the [Gitter][Gitter].

## Install
- clone this repo into a root _/nodezoo_ folder
- run `npm install`

## Running

This micro-service can be ran as part of a complete system or as a single isolated
unit.

### As a complete system
A special system repository is available that runs the complete system using Docker
and Fuge.

- [Nodezoo: The complete system][System]

### Isolated mode
To make testing easier this micro-service can be ran in 'isolated' mode. This mode
allows testing over http using a well defined port. Please note isolated mode means
patterns are not exposed via mesh.

To run in isolated mode:

 - Clone this repository locally,
 - Run `npm install`,
 - Run `npm start isolated`.

A simple http service is supported and can be called using Curl or other Rest clients.
The default port is `8054`. It can be changed using the `COVERALLS_PORT` environment
variable.

## Using with Curl

Any of the messages above can be run using curl in the following format in the command line
```
curl -d '{"role":"coveralls","cmd":"get","name":"YOUR_TEXT_HERE"}' http://localhost:8054/act
```
## Configuration

### Environment Variables
Various settings can be changed using environment variables, see the list below for
all available variable names.

#### COVERALLS_HOST
  - The host to listen on in isolated mode or the host to be published in the Mesh network.
  - Defaults to `localhost` in isolated mode and to `127.0.0.1` in a Mesh network.

#### COVERALLS_PORT
  - The port to listen on in isolated mode.
  - Defaults to `8054`.

#### COVERALLS_REDIS_HOST
  - The host redis listens on.
  - Defaults to `localhost`.

#### COVERALLS_REDIS_PORT
  - The port redis listens on.
  - Defaults to `6379`.

#### COVERALLS_ISOLATED
  - Starts isolated mode.
  - Defaults to `false`.

#### COVERALLS_REGISTRY
  - Changes the Coveralls URL used to retrieve the module info.
  - Defaults to `https://coveralls.io/`.

#### NPM_REGISTRY
  - Changes the npm registry used to retrieve the module info.
  - Defaults to `http://registry.npmjs.org/`.

#### BASE_HOST
  - Mesh Base location.
  - Defaults to `127.0.0.1:39999`.

## Messages Handled

### `role:coveralls,cmd:get`
Returns Coveralls specific data for the module name provided.

```js
seneca.act(`role:coveralls,cmd:get`, {name:'seneca'}, (err, data) => {})
```

### `role:info,req:part`
An alias for `role:coveralls,cmd:get`, allows integration into the wider nodezoo-system.

```js
seneca.act(`role:info,req:part`, {name:'seneca'}, (err, reply) => {})
```

## Messages Emitted

### `role:info,res:part,part:coveralls`

Called in response to a call to `role:info,req:part`.

```js
seneca.add(`role:info,res:part`, (msg, done) => {})
```

## Data Emitted
- name: Name of the module,
- coverageChange: coverage changed,
- coveredPercent: overall coverage percent,
- badgeUrl: Url of the Coveralls badge,
- cached: - The time the data was last cached at.

## Contributing
The [NodeZoo org][] encourages __open__ and __safe__ participation.

- [Code of Conduct][CoC]

If you feel you can help in any way, be it with documentation, examples, extra testing, or new
features please get in touch.

## License
Copyright (c) 2016, Mihai Dima and other contributors.
Licensed under [MIT][].

[BadgeCoveralls]: https://coveralls.io/repos/github/nodezoo/nodezoo-coveralls/badge.svg?branch=master
[BadgeTravis]: https://travis-ci.org/nodezoo/nodezoo-coveralls.svg
[BadgeGitter]: https://badges.gitter.im/Join%20Chat.svg
[BadgeNpm]: https://badge.fury.io/js/nodezoo-coveralls.svg
[Coveralls]: https://coveralls.io/github/nodezoo/nodezoo-coveralls?branch=master
[CoC]: https://github.com/nodezoo/nodezoo-org/blob/master/CoC.md
[Gitter]: https://gitter.im/nodezoo/nodezoo-org
[Travis]: https://travis-ci.org/nodezoo/nodezoo-coveralls
[Logo]: https://raw.githubusercontent.com/nodezoo/nodezoo-org/master/assets/logo-nodezoo.png
[Npm]: http://npmjs.org/nodezoo-coveralls
[Sponsor]: http://nearform.com
[Lead]: https://github.com/mihaidma
[NodeZoo org]: https://github.com/nodezoo
[Lic]: ./LICENSE
[Issue]: https://github.com/nodezoo/nodezoo-coveralls/issues
[Twitter]: http://twitter.com/nodezoo
[System]: https://github.com/nodezoo/nodezoo-system


CHANGELOG.md

## 3.0.0 - July 10, 2016
* Corrected dependency issues
* Fixed package.json info that was missing
* Updated all applicable dependencies
* Verify coveralls.io response header status codes
* Removed toolbag support completely
* Badged and tidied readme
* Added auto reporting to coveralls

## 2.0.0 - June 17, 2016
* Changed the role:info,req:part response
* Enhanced tests

## 1.1.0 - June 13, 2016
* Enhanced the Coveralls info delivered


test/coveralls-data.js
module.exports = {
  'created_at': '2016-06-14T16:08:59.479Z',
  'url': null,
  'commit_message': 'Add npm downloads badge to README',
  'branch': 'master',
  'committer_name': 'GitHub',
  'committer_email': 'noreply@github.com',
  'commit_sha': '3e885275ea6d78225aa79d290dff8f6d6c62f3c3',
  'repo_name': 'senecajs/seneca',
  'badge_url': 'https://s3.amazonaws.com/assets.coveralls.io/badges/coveralls_87.svg',
  'coverage_change': 0.0,
  'covered_percent': 87.2792362768496
}


test/coveralls.test.js
'use strict'

const Code = require('code')
const Lab = require('lab')
const _ = require('lodash')
const Seneca = require('seneca')

const Proxyquire = require('proxyquire')
const NpmSenecaFakeData = require('./npm-data')
const CoverallsSenecaFakeData = require('./coveralls-data')

const lab = exports.lab = Lab.script()
const describe = lab.describe
const before = lab.before
const beforeEach = lab.beforeEach
const it = lab.it
const expect = Code.expect

const NpmRegistry = 'http://registry.npmjs.org/'
const CoverallsUrl = 'https://coveralls.io/'
const invalidPluginName = 'qwerty_qwerty'

var RequestMap = []
const RequestProxy = {
  request: {
    get: (opts, done) => {
      const request = _.find(RequestMap, (o) => { return opts.url.includes(o.urlMatch) })
      if (request) {
        done(request.err, request.response, request.body)
      }
      else {
        done(new Error('invalid request error'), null, null)
      }
    }
  }
}

const Coveralls = Proxyquire('..', RequestProxy)

function createInstance (options) {
  const opts = _.assign({log: 'silent'}, options)
  return Seneca(opts)
  .use('seneca-entity')
  .use(Coveralls, {
    registry: NpmRegistry,
    url: CoverallsUrl
  })
}

process.setMaxListeners(12)

describe('nodezoo-coveralls test suite', () => {
  const si = createInstance()
  const DefaultRequestMap = [
    {
      urlMatch: NpmRegistry + 'seneca',
      err: null,
      response: {},
      body: JSON.stringify(NpmSenecaFakeData)
    },
    {
      urlMatch: 'coveralls.io',
      err: null,
      response: {},
      body: JSON.stringify(CoverallsSenecaFakeData)
    }
  ]

  before((done) => {
    si.ready(function (err) {
      if (err) {
        return process.exit(!console.error(err))
      }
      done()
    })
  })

  beforeEach((done) => {
    RequestMap = DefaultRequestMap.slice()
    done()
  })

  describe('Valid "role:coveralls,cmd:get" calls', () => {
    it('test non cached valid response', (done) => {
      const payload = { 'name': 'seneca' }

      si.act('role:coveralls,cmd:get', payload, function (err, reply) {
        expect(err).to.not.exist()
        expect(reply.ok).to.be.true()
        expect(reply.data.name).to.equal(payload.name)
        expect(reply.data.coveredPercent).to.be.a.number()
        expect(reply.data.coverageChange).to.be.a.number()
        expect(reply.data.badgeUrl).to.exist()
        expect(reply.data.cached).to.exist()

        done(err)
      })
    })

    it('test cached valid response', (done) => {
      const payload = { 'name': 'seneca' }

      si.act('role:coveralls,cmd:get', payload, function (err, reply) {
        expect(err).to.not.exist()

        const cachedOne = reply.data.cached
        expect(cachedOne).to.exist()

        si.act('role:coveralls,cmd:get', payload, (err, reply) => {
          expect(err).to.not.exist()

          const cachedTwo = reply.data.cached
          expect(cachedTwo).to.exist()

          expect(cachedOne).to.equal(cachedTwo)
          done()
        })
      })
    })

    it('test non-cached valid reponse - update flag', (done) => {
      const payload = {name: 'seneca'}

      si.act('role:coveralls,cmd:get', payload, (err, reply) => {
        expect(err).to.not.exist()

        const cachedOne = reply.data.cached
        expect(cachedOne).to.exist()
        payload.update = true

        si.act('role:coveralls,cmd:get', payload, (err, reply) => {
          expect(err).to.not.exist()

          const cachedTwo = reply.data.cached
          expect(cachedTwo).to.exist()

          expect(cachedOne).to.be.below(cachedTwo)

          done()
        })
      })
    })
  })

  describe('Invalid "role:coveralls,cmd:get" calls', () => {
    it('has no error and no data', (done) => {
      const seneca = createInstance()
      const payload = {name: invalidPluginName}
      const invalidPluginMap = {
        urlMatch: invalidPluginName,
        err: null,
        response: {},
        body: '{}'
      }

      seneca.ready(function () {
        RequestMap = _.concat(DefaultRequestMap, invalidPluginMap)

        seneca.act('role:coveralls,cmd:get', payload, (err, reply) => {
          expect(err).to.not.exist()
          expect(reply.ok).to.be.false()
          expect(reply.err).to.exist()

          done()
        })
      })
    })

    it('npm request returns error', (done) => {
      const seneca = createInstance()
      const payload = {name: 'seneca'}
      const errMsg = 'Request failed'
      const failedRequestMap = {
        urlMatch: 'npm',
        err: errMsg,
        response: null,
        body: null
      }

      seneca.ready(function () {
        RequestMap = _.concat([], failedRequestMap)

        seneca.act('role:coveralls,cmd:get', payload, (err, reply) => {
          expect(err).to.not.exist()
          expect(reply.ok).to.be.false()
          expect(reply.err).to.equal(errMsg)

          done()
        })
      })
    })

    it('npm request invalid body', (done) => {
      const seneca = createInstance()
      const payload = {name: 'seneca'}
      const failedRequestMap = {
        urlMatch: 'npm',
        err: null,
        response: null,
        body: {}
      }

      seneca.ready(function () {
        RequestMap = _.concat([], failedRequestMap)

        seneca.act('role:coveralls,cmd:get', payload, (err, reply) => {
          expect(err).to.not.exist()
          expect(reply.ok).to.be.false()
          expect(reply.err).to.exist()

          done()
        })
      })
    })

    it('coveralls request returns error', (done) => {
      const seneca = createInstance()
      const payload = {name: 'seneca'}
      const errMsg = 'Request failed'
      const failedRequestMap = {
        urlMatch: 'coveralls',
        err: errMsg,
        response: null,
        body: null
      }

      seneca.ready(function () {
        RequestMap = _.concat([], failedRequestMap, DefaultRequestMap[0])

        seneca.act('role:coveralls,cmd:get', payload, (err, reply) => {
          expect(err).to.not.exist()
          expect(reply.ok).to.be.false()
          expect(reply.err).to.equal(errMsg)

          done()
        })
      })
    })

    it('coveralls request invalid body', (done) => {
      const seneca = createInstance()
      const payload = {name: 'seneca'}
      const failedRequestMap = {
        urlMatch: 'coveralls',
        err: null,
        response: null,
        body: {}
      }

      seneca.ready(function () {
        RequestMap = _.concat([], failedRequestMap, DefaultRequestMap[0])

        seneca.act('role:coveralls,cmd:get', payload, (err, reply) => {
          expect(err).to.not.exist()
          expect(reply.ok).to.be.false()
          expect(reply.err).to.exist()

          done()
        })
      })
    })

    it('coveralls request invalid body', (done) => {
      const seneca = createInstance()
      const payload = {name: 'seneca'}
      const failedRequestMap = {
        urlMatch: 'coveralls',
        err: null,
        response: { statusCode: 404 },
        body: '<html lang=\'en\'></html>'
      }

      seneca.ready(function () {
        RequestMap = _.concat([], failedRequestMap, DefaultRequestMap[0])

        seneca.act('role:coveralls,cmd:get', payload, (err, reply) => {
          expect(err).to.not.exist()
          expect(reply.ok).to.be.false()
          expect(reply.err).to.exist()

          done()
        })
      })
    })
  })

  describe('Valid "role:info,req:part" calls', () => {
    it('has no error and has data', (done) => {
      const seneca = createInstance()
      const payload = {name: 'seneca'}

      seneca.ready(function () {
        seneca.act('role:info,req:part', payload, (err, reply) => {
          expect(err).to.not.exist()
          expect(reply.ok).to.be.true()

          done()
        })
      })
    })

    it('responds via "role:info,res:part"', (done) => {
      const seneca = createInstance()
      const payload = {name: 'seneca'}

      seneca.ready(function () {
        seneca.add('role:info,res:part', (msg, cb) => {
          expect(msg).to.exist()

          cb()
          done()
        })

        seneca.act('role:info,req:part', payload, (err, reply) => {
          expect(err).to.not.exist()
          expect(reply).to.exist()
          expect(reply.ok).to.be.true()
        })
      })
    })
  })

  describe('Invalid "role:info,req:part" calls', () => {
    it('calls role:coveralls,cmd:get and receives an error', (done) => {
      const seneca = createInstance()
      const payload = {name: 'seneca'}
      const errMsg = 'error'

      seneca.ready(function () {
        seneca.add('role:coveralls,cmd:get', (msg, cb) => {
          expect(msg).to.exist()

          cb(errMsg, null)
          done()
        })

        seneca.act('role:info,req:part', payload, (err, reply) => {
          expect(err).to.not.exist()
          expect(reply).to.exist()
          expect(reply.ok).to.be.false()
          expect(reply.err).to.equal(errMsg)
        })
      })
    })

    it('calls role:coveralls,cmd:get and the npm request fails', (done) => {
      const seneca = createInstance()
      const payload = {name: 'seneca'}
      const errMsg = 'Request failed'
      const failedRequestMap = {
        urlMatch: 'npm',
        err: errMsg,
        response: null,
        body: null
      }

      seneca.ready(function () {
        RequestMap = _.concat([], failedRequestMap)

        seneca.act('role:info,req:part', payload, (err, reply) => {
          expect(err).to.not.exist()
          expect(reply.ok).to.be.false()
          expect(reply.err).to.equal(errMsg)

          done()
        })
      })
    })
  })
})


.toolbagrc.js
'use strict'

const StatsCollector = require('toolbag-plugin-stats-collector')
const UdpReporter = require('toolbag-plugin-udp-reporter')

module.exports = function (defaults, callback) {
  callback(null, {
    plugins: [
      {
        plugin: UdpReporter,
        options: {
          id: 'udp reporter',
          socketType: 'udp4',
          port: 5001,
          host: 'localhost'
        }
      },
      {
        plugin: StatsCollector,
        options: {
          enabled: true,
          period: 1000,
          eventLoopLimit: 30,
          features: {
            process: true,
            system: true,
            cpu: true,
            memory: true,
            gc: false,
            handles: false,
            requests: false,
            eventLoop: true,
            meta: {
              tags: [
                'nodezoo-travis'
              ]
            }
          }
        }
      }
    ]
  })
}


lib/coveralls.js
'use strict'

const Request = require('request')
const GitUrlParse = require('parse-github-url')

let opts = {
}

module.exports = function (options) {
  const seneca = this

  opts = seneca.util.deepextend(opts, options)
  opts.cache = seneca.make$('coveralls_cache')
  opts.cache.load$()

  seneca.add('role:coveralls,cmd:get', cmdGet)
  seneca.add('role:info,req:part', aliasGet)

  return {
    name: 'nodezoo-coveralls'
  }
}

function cmdGet (msg, done) {
  let cache = opts.cache
  let registry = opts.registry + msg.name
  let context = this

  cache.load$(msg.name, (err, cached) => {
    // $lab:coverage:off$
    if (err) {
      context.log.debug(`Cannot load from cache module ${msg.name}, try now to get it remotely`)
    }
    // $lab:coverage:on$

    if (cached && !msg.update) {
      return done(null, {ok: true, data: cached.data$(cached)})
    }

    Request.get({url: registry, gzip: true}, (err, res, body) => {
      if (err) {
        return done(null, {ok: false, err: err})
      }

      let data = null

      try {
        data = JSON.parse(body)
      }
      catch (e) {
        return done(null, {ok: false, err: e})
      }

      const distTags = data['dist-tags'] || {}
      const latest = ((data.versions || {})[distTags.latest]) || {}
      const repository = latest.repository || {}
      const url = repository.url || ''

      const gitInfo = GitUrlParse(url)
      if (!gitInfo) {
        return done(null, {ok: false, err: 'Cannot parse url'})
      }
      gitInfo.cached = cached

      queryCoveralls(gitInfo, done)
    })
  })
}

function queryCoveralls (gitInfo, done) {
  const cache = opts.cache
  const gitUrl = gitInfo.host.slice(0, gitInfo.host.indexOf('.'))
  const registry = opts.url + gitUrl + '/' + gitInfo.repo

  Request.get({url: registry + '.json', gzip: true}, (err, res, body) => {
    if (err) {
      return done(null, {ok: false, err: err})
    }
    if (res && res.statusCode && res.statusCode !== 200) {
      return done(null, {ok: false, err: 'Coverall request not successful'})
    }

    let coverallsData = null

    try {
      coverallsData = JSON.parse(body)
    }
    catch (e) {
      return done(null, {ok: false, err: e})
    }
    const data = {
      name: gitInfo.name,
      url: registry,
      coverageChange: coverallsData.coverage_change,
      coveredPercent: coverallsData.covered_percent,
      badgeUrl: coverallsData.badge_url,
      cached: Date.now()
    }

    function complete (err, entity) {
      if (err) {
        return done(null, {ok: false, err: err})
      }
      else {
        return done(null, {ok: true, data: entity && entity.data$(entity)})
      }
    }

    if (gitInfo.cached) {
      gitInfo.cached.data$(data).save$(complete)
    }
    else {
      data.id$ = gitInfo.name
      cache.make$(data).save$(complete)
    }
  })
}

function aliasGet (msg, done) {
  const seneca = this
  const payload = {name: msg.name}

  seneca.act(`role:coveralls,cmd:get,update:${msg.update || false}`, payload, (err, res) => {
    if (err) {
      return done(null, {ok: false, err: err})
    }

    if (res && res.ok) {
      payload.data = res.data
      seneca.act('role:info,res:part,part:coveralls', payload)
    }

    done(null, res)
  })
}


srv/start.js
'use strict'

const Seneca = require('seneca')
const Entities = require('seneca-entity')
const Mesh = require('seneca-mesh')
const Coveralls = require('../lib/coveralls')
const RedisStore = require('seneca-redis-store')

const envs = process.env
const opts = {
  seneca: {
    tag: envs.COVERALLS_TAG || 'nodezoo-coveralls',
    log: 'none'
  },
  coveralls: {
    registry: envs.NPM_REGISTRY || 'http://registry.npmjs.org/',
    url: envs.COVERALLS_REGISTRY || 'https://coveralls.io/'
  },
  mesh: {
    auto: true,
    host: envs.COVERALLS_HOST || '127.0.0.1',
    bases: [envs.BASE_HOST || '127.0.0.1:39999'],
    listen: [
      {pin: 'role:coveralls,cmd:get', model: 'consume', host: envs.COVERALLS_HOST || '127.0.0.1'},
      {pin: 'role:info,req:part', model: 'observe', host: envs.COVERALLS_HOST || '127.0.0.1'}
    ]
  },
  isolated: {
    host: envs.COVERALLS_HOST || 'localhost',
    port: envs.COVERALLS_PORT || '8054'
  },
  redis: {
    host: envs.COVERALLS_REDIS_HOST || 'localhost',
    port: envs.COVERALLS_REDIS_PORT || '6379'
  }
}

const Service = Seneca(opts.seneca)

Service.use(Entities)

if (envs.COVERALLS_ISOLATED) {
  Service.listen(opts.isolated)
}
else {
  Service.use(Mesh, opts.mesh)
  Service.use(RedisStore, opts.redis)
}

Service.use(Coveralls, opts.coveralls)
